util/ranger/new_refiner.go

```diff
@@ -0,0 +1,271 @@
+		conditions[i] = expression.PushDownNot(cond, false, nil)
+	}
+	for _, cond := range conditions {
+		offset := getEQColOffset(cond, cols)
+		if offset != -1 {
+			accessConds[offset] = cond
+		}
+	}
+	for i, cond := range accessConds {
+		if cond == nil {
+			accessConds = accessConds[:i]
+			accessEqualCount = i
+			break
+		}
+		if lengths[i] != types.UnspecifiedLength {
+			filterConds = append(filterConds, cond)
+		}
+		if i == len(accessConds)-1 {
+			accessEqualCount = len(accessConds)
+		}
+	}
+	accessInAndEqCount = accessEqualCount
+	// We should remove all accessConds, so that they will not be added to filter conditions.
+	conditions = removeAccessConditions(conditions, accessConds)
+	var curIndex int
+	for curIndex = accessEqualCount; curIndex < len(cols); curIndex++ {
+		checker := &conditionChecker{
+			cols:         cols,
+			columnOffset: curIndex,
+			length:       lengths[curIndex],
+		}
+		// First of all, we should extract all of in/eq expressions from rest conditions for every continuous index column.
+		// e.g. For index (a,b,c) and conditions a in (1,2) and b < 1 and c in (3,4), we should only extract column a in (1,2).
+		accessIdx := checker.findEqOrInFunc(conditions)
+		// If we fail to find any in or eq expression, we should consider all of other conditions for the next column.
+		if accessIdx == -1 {
+			accessConds, filterConds = checker.extractAccessAndFilterConds(conditions, accessConds, filterConds)
+			break
+		}
+		accessInAndEqCount++
+		accessConds = append(accessConds, conditions[accessIdx])
+		if lengths[curIndex] != types.UnspecifiedLength {
+			filterConds = append(filterConds, conditions[accessIdx])
+		}
+		conditions = append(conditions[:accessIdx], conditions[accessIdx+1:]...)
+	}
+	// If curIndex equals to len of index columns, it means the rest conditions haven't been appended to filter conditions.
+	if curIndex == len(cols) {
+		filterConds = append(filterConds, conditions...)
+	}
+	return accessConds, filterConds, accessEqualCount, accessInAndEqCount
+}
+
+// buildColumnRange builds the range for sampling histogram to calculate the row count.
+func buildColumnRange(conds []expression.Expression, sc *variable.StatementContext, tp *types.FieldType) ([]*types.ColumnRange, error) {
+	if len(conds) == 0 {
+		return []*types.ColumnRange{{Low: types.Datum{}, High: types.MaxValueDatum()}}, nil
+	}
+
+	rb := Builder{Sc: sc}
+	rangePoints := FullRange
+	for _, cond := range conds {
+		rangePoints = rb.intersection(rangePoints, rb.build(cond))
+		if rb.err != nil {
+			return nil, errors.Trace(rb.err)
+		}
+	}
+	ranges := rb.buildColumnRanges(rangePoints, tp)
+	if rb.err != nil {
+		return nil, errors.Trace(rb.err)
+	}
+	return ranges, nil
+}
+
+// BuildRange is a method which can calculate IntColumnRange, ColumnRange, IndexRange.
+func BuildRange(sc *variable.StatementContext, conds []expression.Expression, rangeType int, cols []*expression.Column, lengths []int) (retRanges []types.Range,
```

Comment:
I think it's no need to pass every column's length. It's enough to know whether the last column has specified length. e.g. if the index is (a,b,c) but b has a length, we can only pass the columns (a,b)